[Ryan Lee] iP

ryanlml opened the pull request:
# Duke

Feel like your life is in a mess? Too many tasks to keep track of? Perhaps you're taking CS2103 this semester?

> "Yesterday is history, tomorrow is a mystery, but today is a gift. That's why it is called the present" - Master Oogway [(Source)](https://www.goodreads.com/author/quotes/7178454.Master_Oogway)

Fret not, Duke is an interactive chatbot that helps schedule your time so that you can focus on things that truly matters. Duke is

- Text-based
- Easy to learn
- ~~Fast~~ ***SUPER FAST*** to use!

All you need to do is,

1. Download it here
2. Double click it
3. Add tasks
4. Done!

FEATURES:

- [x] Add and delete tasks
- [x] Find tasks by deadline
- [ ] Undo and redo features

Duke supports the following types of tasks:

```
deadline
event
todo
```

Duke is written in Java, so you may be familiar with

```Java
public static void main(String[] args) {
    System.out.println("Hello World!");
}
```

Be a master of your time, use Duke today! :+1:

Commit:
Add Gradle support

Commit:
added Level 1. Greet, Echo, Exit

ryanlml committed:
added Level 2. Add, List

ryanlml committed:
Added Level 3. Mark as Done

ryanlml committed:
Added Level 4. ToDos, Events, Deadlines

ryanlml committed:
Added Level 5. Handle Errors

ryanlml committed:
Added Level 6. Delete

ryanlml committed:
- Added more exception handling, where users may input invalid tasks to mark, unmark or delete.
- Cleaned up the code

ryanlml committed:
Added Level 7. Save and added methods in Task to streamline saving and loading from file.

ryanlml committed:
Added Level 8. Dates and Times and a command deadlines to see events/deadlines on a specific date.

ryanlml committed:
Merge branch 'master' into branch-Level-8

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

ryanlml committed:
Deleted unnecessary files

ryanlml committed:
Added A-MoreOOP

ryanlml committed:
Added A-Packages.

ryanlml committed:
Bug fixes

ryanlml committed:
Added A-Gradle

ryanlml committed:
Added A-JUnit.

ryanlml committed:
Fixed JUnit issues

ryanlml committed:
Added A-jar.

ryanlml committed:
Added TextUiTesting.

ryanlml committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

# Conflicts:
#	build.gradle

ryanlml committed:
Added A-JavaDoc.

ryanlml committed:
Added A-CodingStandard.

ryanlml committed:
Added Level-9.

ryanlml committed:
Merge branch 'branch-A-JavaDoc'

ryanlml committed:
Merge branch 'branch-Level-9'

jhchee18 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,52 @@
+        return this.isDone;
+    }
+
+    public void mark() {
+        this.isDone = true;
+    }
+
+    public void unmark() {
+        this.isDone = false;
+    }
```
I think if to preserve the immutability concept taught in CS2030, `mark()` and `unnmark()` should create a new Task instead of changing its variables.

jhchee18 commented on src/main/java/duke/TripleConsumer.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+
+public interface TripleConsumer {
+    /**
+     * Interface for our command.
+     *
+     * @param tasks
+     * @param ui
+     * @param storage
+     */
+    void execute(TaskList tasks, Ui ui, Storage storage);
+}
```
I like how you package the three params into TripleConsumer and apply it by using Stream. Learned from you :) Thanks.

jhchee18 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,270 @@
+
+public class Parser {
+
+    private static final String TIME_INPUT_PATTERN = "dd-MM-yyyy HHmm";
+    private static final String DEADLINE_TIME_INPUT_PATTERN = "dd-MM-yyyy";
```
Good code quality practice as to avoid magic numbers. I like how you always keep things clear and simple.

jhchee18 reviewed (commented):
Excellent code quality, made me hard to find places to comment on 😄 I learned a lot from your code too. Good job!

ryanlml committed:
Added Level-10: GUI

gqsg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,270 @@
+     * Parser to interpret the commands.
+     *
+     * @param fullCommand
+     * @return
```
Did you forget to add description for @return?

gqsg commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Handles all system output messages.
+     */
+
```
Perhaps you can remove blank line between the documentation block and the method/class to comply with the coding standard.

gqsg reviewed (commented):
Good Job! No major issues but perhaps you can consider to divide the classes into more suitable java packages, for example, put Ui.java in a package called ui, Event.java, Deadline.java and Todo.java into a package called task.

snigloo commented on src/main/java/duke/Command.java:
```diff
@@ -5,6 +5,12 @@ public class Command {
+    /**
+     * A command class to keep track of commands.
+     *
+     * @param function
```
For your JavaDocs, please describe the parameter to be input into each method.

snigloo commented on src/main/java/duke/Deadline.java:
```diff
@@ -6,6 +6,13 @@ public class Deadline extends Task {
 
     protected LocalDateTime by;
```
Should your variable be just "by"? It may not be clear.

snigloo commented on src/main/java/duke/TaskList.java:
```diff
@@ -6,6 +6,10 @@
+
     protected ArrayList<Task> tasks;
 
     public TaskList() {
```
Make sure to add JavaDocs to most of your methods.

snigloo commented on src/main/java/duke/Storage.java:
```diff
@@ -53,6 +61,13 @@ public ArrayList<Task> load() throws DukeException, FileNotFoundException {
         return tasks;
```
Note that switch does not require indentation.

snigloo commented on src/main/java/duke/Storage.java:
```diff
@@ -18,6 +18,14 @@ public Storage(String filePath) {
+
     public ArrayList<Task> load() throws DukeException, FileNotFoundException {
 
         ArrayList<Task> tasks = new ArrayList<Task>();
```
Could process some of these commands in Parser.

snigloo commented on src/main/java/duke/Parser.java:
```diff
@@ -10,6 +10,14 @@ public class Parser {
+
     public static Command parse(String fullCommand) throws DukeException {
 
         String instruction = fullCommand.contains(" ") ? fullCommand.split(" ", 2)[0] : fullCommand;
```
Try to avoid deep nesting (more than 3 indentations) in your parse method. :)

snigloo reviewed (commented):
Overall, generally great code. Great use of abstraction and OOP, with just minor fixes to deep nesting and coding standard.

ryanlml committed:
Fixed some issues with GUI

ryanlml committed:
Added A-Assertions.

ryanlml committed:
Merge pull request #1 from ryanlml/branch-A-Assertions

Added A-Assertions.

ryanlml committed:
Added A-CodeQuality.

Fixed the following coding standard violations:
- Missing @parm descriptions in JavaDocs
- Missing @return descriptions in JavaDocs
- Removed all extra line breaks between JavaDocs and the method
- Removed unnecessary line breaks to improve code readability
- Corrected wrong indentation between switch and case statements

Thanks to @jhchee18, @NUSDG and @snigloo for pointing these out!

ryanlml committed:
Fixed more coding standard violations.

ryanlml committed:
Added C-Tagging

ryanlml committed:
Improved the tagging system.
- You can now add multiple tags to a task.
- You can search for tasks with a particular tag.
- You can add / remove tags on any task.

Also abstracted out repeated code, made a method showTasksWithThisProperty to easily filter and print tasks.
Pushed (most) system.out.println() to the Ui class.

ryanlml committed:
Already using streams where appropriate

ryanlml committed:
Packaged tasks into its own file.

ryanlml committed:
Packaged java classes appropriately to adhere to SLAP.
Updated images for Duke and User.
Completed a User Guide.
Improvements to the GUI, no text should be hidden in ellipses anymore.

ryanlml committed:
Updated User Guide

ryanlml committed:
Update README.md

ryanlml committed:
Updated User Guide

ryanlml committed:
Merge commit '61d835d86c6d2a9de0522f986559bdb6ef7b896a'

ryanlml committed:
Updated README.md

ryanlml committed:
Cleaned up code for final release.
- Removed unused imports
- Added more JavaDocs

ryanlml committed:
Update README.md

ryanlml committed:
Update README.md
